File: algorithms/js/arrays/hashing/subarraySumEqualsK.js
// Given an array of integers nums and an integer k, 
// return the total number of subarrays whose sum equals to k.

// A subarray is a contiguous non-empty sequence of elements within an array.

// Pattern: Subarray sum → Prefix Sum + HashMap

// Invariant:
// The hashmap stores how many times each prefix sum has appeared so far.

// Action:
// As we iterate through the array we maintain a runningSum.
// If runningSum - k has appeared before, then a subarray ending here sums to k.
// We add the frequency of (runningSum - k) to the count.
// Then we record the current runningSum in the hashmap.

// Why:
// prefixSum[i] - prefixSum[j] = k
// Therefore prefixSum[j] = prefixSum[i] - k
// If we've seen prefixSum[i] - k before, that means a valid subarray exists.

const subarraySum = (nums, k) => {
    let count = 0;
    let runningSum = 0;
    const sumMap = new Map();
    sumMap.set(runningSum, 1);

    for (let num of nums) {
        // CurrentTotalSum - previousRunningSum = k
        // -previousRunningSum = k - CurrentTotalSum
        // previousRunningSum = -k + CurrentTotalSum

        runningSum += num;
        if (sumMap.has(runningSum - k)) {
            count += sumMap.get(runningSum - k);
        }

        sumMap.set(runningSum, (sumMap.get(runningSum) || 0) + 1);
    }

    return count;
}

File: algorithms/py/arrays/hashing/subarraySumEqualsK.py
# Given an array of integers nums and an integer k, return the total number of subarrays whose sum equals to k.
# A subarray is a contiguous non-empty sequence of elements within an array.

# Pattern:
#     This is a prefix sum with hashmap problem

# Approach:
#     I keep track of the runningSum in a hashmap as I iterate through the array of integers.
#     If currentRunningSum - previousRunningSum = k then we have a subarray where the sum equals to k
#     To find the previousRunningSum we compute currentRunningSum - k = previousRunningSum and check for the previousRunningSum in hashmap
#     We keep track of the frequency of runningSum in the hashmap because the same runningSum can appear multiple times and each occurrence
#     represents a different starting point for a valid subarray

# Time & Space complexity:
#     Time complexity of this is linear O(n)
#     Space complexity is O(n) for the hashmap we use to keep track of runningSums

# EdgeCase(s):
#     If the first number matches k then we need to be able to find currentRunningSum - k = 0 in hashmap



class Solution:
    def subArraySum(self, nums: list[int], k: int) -> int:
        hashmap = {0: 1}
        runningSum = 0
        count = 0

        for i, num in enumerate(nums):
            # currentSum - previousSum = k
            # currentSum - k = previousSum
            runningSum += num

            if runningSum - k in hashmap:
                count += hashmap[runningSum - k]

            hashmap[runningSum] = hashmap.get(runningSum, 0) + 1

        return count